src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,293 @@
+package com.android.volley.cronet;
+
+import android.content.Context;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.android.volley.AuthFailureError;
+import com.android.volley.Header;
+import com.android.volley.Request;
+import com.android.volley.toolbox.AsyncHttpStack;
+import com.android.volley.toolbox.ByteArrayPool;
+import com.android.volley.toolbox.HttpResponse;
+import com.android.volley.toolbox.PoolingByteArrayOutputStream;
+import com.android.volley.toolbox.UrlRewriter;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import org.chromium.net.CronetEngine;
+import org.chromium.net.CronetException;
+import org.chromium.net.UploadDataProvider;
+import org.chromium.net.UploadDataProviders;
+import org.chromium.net.UrlRequest;
+import org.chromium.net.UrlRequest.Callback;
+import org.chromium.net.UrlResponseInfo;
+
+/**
+ * A {@link AsyncHttpStack} that's based on Cronet's fully asynchronous API for network requests.
+ */
+public class CronetHttpStack extends AsyncHttpStack {
+
+    private final CronetEngine mCronetEngine;
+    private final Executor mCallbackExecutor;
+    private final ExecutorService mBlockingExecutor;
+    private final ByteArrayPool mPool;
+    private final UrlRewriter mUrlRewriter;
+
+    private CronetHttpStack(Builder builder) {
```

Comment:
nit: My recommendation would be to have the constructor take the arguments directly that it expects, and have the Builder call the constructor with those arguments.

This is the more conventional pattern that I've seen for this sort of thing, and the benefit of that approach is that it makes unit testing the class simpler as you don't need to use the builder interface just to create a test instance.

Reply:
Okay, that makes sense. I saw two different ways to do it, but I'll change it.